apps/api/src/app/workflows-v2/util/template-parser/new-liquid-parser.ts

```diff
@@ -0,0 +1,947 @@
+    if (localVariables.has(variableName) || !variableSchema || variableSchema.additionalProperties === true) {
+      validVariables.push({
+        name: variableName,
+        output,
+        outputStart,
+        outputEnd,
+      });
+      return;
+    }
+
+    // Otherwise, it's invalid (missing namespace)
+    invalidVariables.push({
+      name: variableName,
+      message: `missing namespace. Did you mean {{payload.${variableName}}}?`,
+      output,
+      outputStart,
+      outputEnd,
+    });
+    return;
+  }
+
+  // Variable has namespace, proceed with normal validation
+  const isLocalVariable = Array.from(localVariables).some(
+    (localVar) => variableName === localVar || variableName.startsWith(localVar + '.')
+  );
+
+  if (isLocalVariable) {
+    validVariables.push({
+      name: variableName,
+      output,
+      outputStart,
+      outputEnd,
+    });
+  } else {
+    const isAllowedVariable = isPropertyAllowed(variableSchema, variableName);
+    if (isAllowedVariable) {
+      validVariables.push({
+        name: variableName,
+        output,
+        outputStart,
+        outputEnd,
+      });
+
+      // Handle filter arguments (like toSentence)
+      if (filters.length > 0) {
+        filters.forEach((filter) => {
+          const { args } = filter;
+          const firstArg = args[0];
+          if (
+            filter.name === 'toSentence' &&
+            args.length > 0 &&
+            'content' in firstArg &&
+            typeof firstArg.content === 'string'
+          ) {
+            /**
+             * Check if the parent variable with the first argument is allowed
+             * basically forcing it to check if additionalProperties is true by checking for final variable name
+             * and if the parent variable is a valid dynamic path as variableSchema can be undefined.
+             * OR
+             * Check if the variable is a digest events array variable
+             * and the first argument starts with payload.
+             */
+            if (
+              (isValidDynamicPath(variableName) &&
+                isPropertyAllowed(variableSchema, `${variableName}.${firstArg.content}`)) ||
+              (firstArg.content.startsWith('payload.') && DIGEST_EVENTS_VARIABLE_PATTERN.test(variableName))
+            ) {
+              const isFirstArgValid = isPropertyAllowed(variableSchema, firstArg.content);
+              if (isFirstArgValid) {
+                validVariables.push({
+                  name: `${variableName}.${firstArg.content}`,
+                  output: firstArg.content,
+                  outputStart,
+                  outputEnd,
+                });
+              }
+            }
+          }
+        });
+      }
+    } else {
+      invalidVariables.push({
+        name: variableName,
+        message: 'is not supported',
+        output,
+        outputStart,
+        outputEnd,
+      });
+    }
+  }
+}
+
+function processTagToken({
```

Comment:
process different liquid tags, for example for loops, conditions, assignments etc